# Data Merging Basics

Learn how you can merge disparate data using inner joins. By combining information from multiple sources you’ll uncover compelling insights that may have previously been hidden. You’ll also learn how the relationship between those sources, such as one-to-one or one-to-many, can affect your result.


In [1]:
import pandas as pd

# Example DataFrames
df1 = pd.DataFrame({
    'ID': [1, 2, 3],
    'Name': ['Alice', 'Bob', 'Charlie'],
    'Value': [100, 200, 300]
})

df2 = pd.DataFrame({
    'ID': [2, 3, 4],
    'Value': [85, 90, 95]
})


## Inner Join

An Inner join will only return rows that have matching values in both tables.

- Tables = **DataFrames**
- Merging = **Joining**

### **Syntax:**

```python
result = left_dataframe.merge(right_dataframe, on='column_name', how='join_type')
```

#### **Parameters:**

* `left_dataframe`: The first DataFrame.
* `right_dataframe`: The second DataFrame.
* `on='column_name'`: The common column(s) to merge on (must exist in both DataFrames).
* `how='join_type'`: Type of join (`'inner'`,`'left'`,`'right'`,`'outer'`).
* default join is **inner**

```python
import pandas as pd

df1 = pd.DataFrame({
    'ID': [1, 2, 3],
    'Name': ['Alice', 'Bob', 'Charlie']
})

df2 = pd.DataFrame({
    'ID': [2, 3, 4],
    'Score': [85, 90, 95]
})

result = pd.merge(df1, df2, on='ID', how='inner')

print(result)
```

| ID  | Name   | Score |
| ----- | -------- | ------- |
| 0 2 | Bob    | 85    |
| 1 3 | Charle | 90    |



In [2]:
result = df1.merge(df2, on='ID', how='inner')
print(result)

   ID     Name  Value_x  Value_y
0   2      Bob      200       85
1   3  Charlie      300       90


### Suffixes
The suffixes are only applied to overlapping column names that are not used in the on parameter.

```python
result = df1.merge(df2, on='ID', suffixes=('_info', '_score'))
```

In [3]:
result = df1.merge(df2, on='ID', suffixes=('_df1', '_df2'))
print(result)

   ID     Name  Value_df1  Value_df2
0   2      Bob        200         85
1   3  Charlie        300         90


# What column to merge on?

Chicago provides a list of taxicab owners and vehicles licensed to operate within the city, for public safety. Your goal is to merge two tables together. One table is called `taxi_owners`, with info about the taxi cab company owners, and one is called `taxi_veh`, with info about each taxi cab vehicle. Both the `taxi_owners` and `taxi_veh` tables have been loaded for you to explore.

Choose the column you would use to merge the two tables on using the `.merge()` method.

***Possible answers***

- `on='rid'`

- `on='vid'` Correct Answer

- `on='year'`

- `on='zip'`

Yes, great job! Both DataFrames contained the column `vid`. Now continue on to the next exercise where you will using this information to merge the tables.


### Your first inner join

You have been tasked with figuring out what the most popular types of fuel used in Chicago taxis are. To complete the analysis, you need to merge the `taxi_owners` and `taxi_veh` tables together on the `vid` column. You can then use the merged table along with the `.value_counts()` method to find the most common `fuel_type`.

Since you'll be working with `pandas` throughout the course, the package will be preloaded for you as `pd` in each exercise in this course. Also the `taxi_owners` and `taxi_veh` DataFrames are loaded for you.

#### Instructions 1/3       **35 XP**

Merge `taxi_owners` with `taxi_veh` on the column `vid`, and save the result to `taxi_own_veh`.


In [4]:
taxi_owners = pd.read_pickle("../data/taxi_owners.p")
taxi_veh = pd.read_pickle("../data/taxi_vehicles.p")

print(taxi_owners.head())
print(taxi_veh.head())

     rid   vid           owner                 address    zip
0  T6285  6285  AGEAN TAXI LLC     4536 N. ELSTON AVE.  60630
1  T4862  4862    MANGIB CORP.  5717 N. WASHTENAW AVE.  60659
2  T1495  1495   FUNRIDE, INC.     3351 W. ADDISON ST.  60618
3  T4231  4231    ALQUSH CORP.   6611 N. CAMPBELL AVE.  60645
4  T5971  5971  EUNIFFORD INC.     3351 W. ADDISON ST.  60618
    vid    make   model  year fuel_type                owner
0  2767  TOYOTA   CAMRY  2013    HYBRID       SEYED M. BADRI
1  1411  TOYOTA    RAV4  2017    HYBRID          DESZY CORP.
2  6500  NISSAN  SENTRA  2019  GASOLINE       AGAPH CAB CORP
3  2746  TOYOTA   CAMRY  2013    HYBRID  MIDWEST CAB CO, INC
4  5922  TOYOTA   CAMRY  2013    HYBRID       SUMETTI CAB CO


In [5]:
taxi_own_veh = taxi_owners.merge(taxi_veh, on="vid")

In [6]:
print(taxi_own_veh.head())
print(taxi_own_veh.columns)

     rid   vid         owner_x                 address    zip    make   model  \
0  T6285  6285  AGEAN TAXI LLC     4536 N. ELSTON AVE.  60630  NISSAN  ALTIMA   
1  T4862  4862    MANGIB CORP.  5717 N. WASHTENAW AVE.  60659   HONDA     CRV   
2  T1495  1495   FUNRIDE, INC.     3351 W. ADDISON ST.  60618  TOYOTA  SIENNA   
3  T4231  4231    ALQUSH CORP.   6611 N. CAMPBELL AVE.  60645  TOYOTA   CAMRY   
4  T5971  5971  EUNIFFORD INC.     3351 W. ADDISON ST.  60618  TOYOTA  SIENNA   

   year fuel_type         owner_y  
0  2011    HYBRID  AGEAN TAXI LLC  
1  2014  GASOLINE    MANGIB CORP.  
2  2015  GASOLINE   FUNRIDE, INC.  
3  2014    HYBRID    ALQUSH CORP.  
4  2015  GASOLINE  EUNIFFORD INC.  
Index(['rid', 'vid', 'owner_x', 'address', 'zip', 'make', 'model', 'year',
       'fuel_type', 'owner_y'],
      dtype='object')


#### Instructions 2/3
Set the left and right table suffixes for overlapping columns of the merge to `_own` and `_veh`, respectively.

In [7]:
# Merge the taxi_owners and taxi_veh tables setting a suffix
taxi_own_veh = taxi_owners.merge(taxi_veh, on='vid', suffixes=('_own', '_veh'))

# Print the column names of taxi_own_veh
print(taxi_own_veh.columns)

Index(['rid', 'vid', 'owner_own', 'address', 'zip', 'make', 'model', 'year',
       'fuel_type', 'owner_veh'],
      dtype='object')


#### Instruction 3/3

Select the `fuel_type` column from `taxi_own_veh` and print the `value_counts()` to find the most popular `fuel_type`s used.


In [8]:
# Merge the taxi_owners and taxi_veh tables setting a suffix
taxi_own_veh = taxi_owners.merge(taxi_veh, on='vid', suffixes=('_own','_veh'))

# Print the value_counts to find the most popular fuel_type
print(taxi_own_veh['fuel_type'].value_counts())

fuel_type
HYBRID                    2792
GASOLINE                   611
FLEX FUEL                   89
COMPRESSED NATURAL GAS      27
Name: count, dtype: int64


# Inner joins and number of rows returned

All of the merges you have studied to this point are called inner joins. It is necessary to understand that inner joins only return the rows with matching values in both tables. You will explore this further by reviewing the merge between the `wards` and `census` tables, then comparing it to merges of copies of these tables that are slightly altered, named `wards_altered`, and `census_altered`. The first row of the `wards` column has been changed in the altered tables. You will examine how this affects the merge between them. The tables have been loaded for you.

For this exercise, it is important to know that the `wards` and `census` tables start with **50** rows.


## Instructions 1/3     **35 XP**

Merge `wards` and `census` on the `ward` column and save the result to `wards_census`.


In [11]:
wards = pd.read_pickle("../data/ward.p")
census = pd.read_pickle("../data/census.p")

In [12]:
# Merge the wards and census tables on the ward column
wards_census = wards.merge(census, on='ward')

# Print the shape of wards_census
print('wards_census table shape:', wards_census.shape)

wards_census table shape: (50, 9)


#### Instruction 2/3
Merge the `wards_altered` and `census` tables on the `ward` column, and notice the difference in returned rows


In [13]:
wards_altered = wards
census_altered = census
# Print the first few rows of the wards_altered table to view the change
print(wards_altered[['ward']].head())

# Merge the wards_altered and census tables on the ward column
wards_altered_census = wards_altered.merge(census, on='ward')

# Print the shape of wards_altered_census
print('wards_altered_census table shape:', wards_altered_census.shape)

  ward
0    1
1    2
2    3
3    4
4    5
wards_altered_census table shape: (50, 9)


#### Instructions 3/3

Merge the `wards` and `census_altered` tables on the `ward` column, and notice the difference in returned rows.


In [14]:
# Print the first few rows of the census_altered table to view the change
print(census_altered[['ward']].head())

# Merge the wards and census_altered tables on the ward column
wards_census_altered = wards.merge(census_altered, on='ward')

# Print the shape of wards_census_altered
print('wards_census_altered table shape:', wards_census_altered.shape)

  ward
0    1
1    2
2    3
3    4
4    5
wards_census_altered table shape: (50, 9)


Great job! In step 1, the `.merge()` returned a table with the same number of rows as the original `wards` table. However, in steps 2 and 3, using the altered tables with the altered first row of the `ward` column, the number of returned rows was fewer. There was not a matching value in the `ward` column of the other table. _Remember that `.merge()` only returns rows where the values match in both tables._


## One-to-One Relationships

![1:1](https://i.ibb.co/WGbT565/Screenshot-2024-12-04-094133.png)

Every row in the left table is related to only one row in the right table

### One-to-Many Relationships

Every row in left table is related to one or more rows in the right table

![1:M](https://i.ibb.co/0ZhBS4L/Screenshot-2024-12-04-094918.png)

### One-to-many classification

Understanding the difference between a one-to-one and one-to-many relationship is a useful skill. In this exercise, consider a set of tables from an e-commerce website. The hypothetical tables are the following:

* A `customer` table with information about each customer
* A `cust_tax_info` table with customers unique tax IDs
* An `orders` table with information about each order
* A `products` table with details about each unique product sold
* An `inventory` table with information on how much total inventory is available to sell for each product

#### **Instructions**       **100XP**

* Select the relationship type that is most appropriate for the relationship between the different tables: **One-to-one** , or **One-to-many** .

**One-to-One**
- The relationship between `customer` and `cust_tax_info` .
- The relationship between `products` and `inventory` .

**One-to-Many**
- The relationship between the `products` and `orders` .
- The relationship between the `customers` and `orders` .



# One-to-many merge

A business may have one or multiple owners. In this exercise, you will continue to gain experience with one-to-many merges by merging a table of business owners, called `biz_owners`, to the `licenses` table. Recall from the video lesson, with a one-to-many relationship, a row in the left table may be repeated if it is related to multiple rows in the right table. In this lesson, you will explore this further by finding out what is the most common business owner title. (i.e., secretary, CEO, or vice president)

The `licenses` and `biz_owners` DataFrames are loaded for you.

## Instructions         **100 XP**

* Starting with the `licenses` table on the left, merge it to the `biz_owners` table on the column `account`, and save the results to a variable named `licenses_owners`.
* Group `licenses_owners` by `title` and count the number of accounts for each title. Save the result as `counted_df`
* Sort `counted_df` by the number of **accounts** in **descending order** , and save this as a variable named `sorted_df`.
* Use the `.head()` method to print the first few rows of the `sorted_df`.


In [18]:
licenses = pd.read_pickle("../data/licenses.p")
biz_owners = pd.read_pickle("../data/business_owners.p")

# Merge the licenses and biz_owners table on account
licenses_owners =licenses.merge(biz_owners, on='account')

# Group the results by title then count the number of accounts
counted_df = licenses_owners.groupby("title").agg({'account':'count'})

# Sort the counted_df in descending order
sorted_df = counted_df.sort_values("account", ascending=False)

# Use .head() method to print the first few rows of sorted_df
print(sorted_df.head())

                 account
title                   
PRESIDENT           6259
SECRETARY           5205
SOLE PROPRIETOR     1658
OTHER               1200
VICE PRESIDENT       970


Wonderful! After merging the tables together, you counted the number of repeated rows with the combination of `.groupby()` and `.agg()` statements. You see that president, followed by secretary, are the most common business owner titles.


## Merging Multiple DataFrames
![Multiple Merge](https://i.ibb.co/DKJ3GMt/Screenshot-2024-12-04-133609.png)

### Single Merge
```python
letf_table.merge(right_table, on=['col_1', 'col_2'])
```
```python
grants.merge(licenses, on=['address', 'zip'])
```
* **Start with the primary DataFrame** :`grants`
* **Use `.merge()`** : to combine data.
* **Add the second DataFrame as the first argument** :`licenses`.
* **Specify the common columns with `on=[...]`** :`['address', 'zip']`

* Pass a List of column names we want to marge on to the 'on' argument.

### Multiple Table Merge
```python
FirstDF.merge(SecondDF, on=[common_columns], suffixes=(suffix1, suffix2))\
   .merge(ThirdDF, on=[common_column], suffixes=(suffix1, suffix2))
```

```python
grant_licenses_ward = grants.merge(licenses, on=['address', 'zip'])\
    .merge(wards, on='ward', suffixes=('_bus', '_ward'))
```

---------

1. **Start with the base DataFrame** :`grants`.
2. **Use `.merge()` to add the first table** :`licenses` (with common columns`['address', 'zip']`).
3. **Chain `.merge()` to add the next table** :`wards` (with the common column`'ward'`).
4. **Add `suffixes` if column names overlap** :`('_bus', '_ward')`.


#### **Step-by-Step Thinking:**

1. First,**combine `grants` with `licenses`** : Use the common columns`['address', 'zip']`.
2. Next,**combine the resulting DataFrame with `wards`** : Use the common column`'ward'`.
3. Add**suffixes to handle overlapping column names** :`('_bus', '_ward')`.

This makes the logic of chaining merges easy to grasp: **"Base + Add-On1 + Add-On2"** .


### Results
```python
import matplotlib.pyplot as plt

grant_licenses_ward.groupby('ward').agg('sum').plot(kind='bar', y='grant')
plt.show()
```

### Merging even more...

### Three tables:
```python
df1.merge(df2, on='col') \
    .merge(df3, on='col')
```

### Four tables:
```python
df1.merge(df2, on='col') \
    .merge(df3, on='col') \
    .merge(df4, on='col')
```




### Total riders in a month

Your goal is to find the total number of rides provided to passengers passing through the Wilson station (`station_name == 'Wilson'`) when riding Chicago's public transportation system on weekdays (`day_type == 'Weekday'`) in July (`month == 7`). Luckily, Chicago provides this detailed data, but it is in three different tables. You will work on merging these tables together to answer the question. This data is different from the business related data you have seen so far, but all the information you need to answer the question is provided.

The `cal`, `ridership`, and `stations` DataFrames have been loaded for you. The relationship between the tables can be seen in the diagram below.

![Table diagram. The cal table relates to ridership via year, month, and day. The ridership table relates to the stations table via station_id.](https://assets.datacamp.com/production/repositories/5486/datasets/56b5ecb2edcdc896c69effdf05ef65e5454ff996/cta_L_diagram.png)

#### Instructions 1/3             **35 XP**

* Merge the `ridership` and `cal` tables together, starting with the `ridership` table on the left and save the result to the variable `ridership_cal`. If you code takes too long to run, your merge conditions might be incorrect.


In [30]:
cal  = pd.read_pickle("../data/cta_calendar.p")
ridership = pd.read_pickle("../data/cta_ridership.p")
stations = pd.read_pickle("../data/stations.p")


print("----------------------------")
print(cal.columns)
print(ridership.columns)
print(stations.columns)
print("----------------------------")

# Merge the ridership and cal tables
ridership_cal = ridership.merge(cal, on=['year', 'month', 'day'])
print(ridership_cal.head())

----------------------------
Index(['year', 'month', 'day', 'day_type'], dtype='object')
Index(['station_id', 'year', 'month', 'day', 'rides'], dtype='object')
Index(['station_id', 'station_name', 'location'], dtype='object')
----------------------------
  station_id  year  month  day  rides        day_type
0      40010  2019      1    1    576  Sunday/Holiday
1      40010  2019      1    2   1457         Weekday
2      40010  2019      1    3   1543         Weekday
3      40010  2019      1    4   1621         Weekday
4      40010  2019      1    5    719        Saturday


#### Instruction 2/3

* Extend the previous merge to three tables by also merging the `stations` table.

In [29]:
# Merge the ridership, cal, and stations tables
ridership_cal_stations = ridership.merge(cal, on=['year','month','day']) \
    .merge(stations, on="station_id")

ridership_cal_stations.head()

station_id  year  month  day  rides        day_type        station_name  \
0      40010  2019      1    1    576  Sunday/Holiday  Austin-Forest Park   
1      40010  2019      1    2   1457         Weekday  Austin-Forest Park   
2      40010  2019      1    3   1543         Weekday  Austin-Forest Park   
3      40010  2019      1    4   1621         Weekday  Austin-Forest Park   
4      40010  2019      1    5    719        Saturday  Austin-Forest Park   

                  location  
0  (41.870851, -87.776812)  
1  (41.870851, -87.776812)  
2  (41.870851, -87.776812)  
3  (41.870851, -87.776812)  
4  (41.870851, -87.776812)

#### Instructions
* Create a variable called `filter_criteria` to select the appropriate rows from the merged table so that you can sum the `rides` column.

In [32]:
# Merge the ridership, cal, and stations tables
ridership_cal_stations = ridership.merge(cal, on=['year','month','day']) \
							.merge(stations, on='station_id')

# Create a filter to filter ridership_cal_stations
filter_criteria = ((ridership_cal_stations['month'] == 7)
                   & (ridership_cal_stations['day_type'] == 'Weekday')
                   & (ridership_cal_stations['station_name'] == 'Wilson'))

# Use .loc and the filter to select for rides
print(ridership_cal_stations.loc[filter_criteria, 'rides'].sum())


140005


Awesome work! You merged three DataFrames together, including merging two tables on multiple columns. Once the tables were merged, you filtered and selected just like any other DataFrame. Finally, you found out that the Wilson station had 140,005 riders during weekdays in July.

### Three table merge

To solidify the concept of a three DataFrame merge, practice another exercise. A reasonable extension of our review of Chicago business data would include looking at demographics information about the neighborhoods where the businesses are. A table with the median income by zip code has been provided to you. You will merge the `licenses` and `wards` tables with this new income-by-zip-code table called `zip_demo`.

The `licenses`, `wards`, and `zip_demo` DataFrames have been loaded for you.

#### Instructions     **100 XP**

* Starting with the `licenses` table, merge to it the `zip_demo` table on the `zip` column. Then merge the resulting table to the `wards` table on the `ward` column. Save result of the three merged tables to a variable named `licenses_zip_ward`.
* Group the results of the three merged tables by the column `alderman` and find the median `income`.


In [38]:
licenses = pd.read_pickle("../data/licenses.p")
zip_demo = pd.read_pickle("../data/zip_demo.p")
wards = pd.read_pickle("../data/ward.p")
# Merge licenses and zip_demo, on zip; and merge the wards on ward
licenses_zip_ward = licenses.merge(zip_demo, on='zip') \
    .merge(wards, on='ward')

# licenses_zip_ward.head()

# Print the results by alderman and show median income
print(licenses_zip_ward.groupby("alderman").agg({'income':'median'}))

                             income
alderman                           
Ameya Pawar                 66246.0
Anthony A. Beale            38206.0
Anthony V. Napolitano       82226.0
Ariel E. Reyboras           41307.0
Brendan Reilly             110215.0
Brian Hopkins               87143.0
Carlos Ramirez-Rosa         66246.0
Carrie M. Austin            38206.0
Chris Taliaferro            55566.0
Daniel "Danny" Solis        41226.0
David H. Moore              33304.0
Deborah Mell                66246.0
Debra L. Silverstein        50554.0
Derrick G. Curtis           65770.0
Edward M. Burke             42335.0
Emma M. Mitts               36283.0
George Cardenas             33959.0
Gilbert Villegas            41307.0
Gregory I. Mitchell         24941.0
Harry Osterman              45442.0
Howard B. Brookins, Jr.     33304.0
James Cappleman             79565.0
Jason C. Ervin              41226.0
Joe Moore                   39163.0
John S. Arena               70122.0
Leslie A. Hairston          

### One-to-many merge with multiple tables

In this exercise, assume that you are looking to start a business in the city of Chicago. Your perfect idea is to start a company that uses goats to mow the lawn for other businesses. However, you have to choose a location in the city to put your goat farm. You need a location with a great deal of space and relatively few businesses and people around to avoid complaints about the smell. You will need to merge three tables to help you choose your location. The `land_use` table has info on the percentage of vacant land by city ward. The `census` table has population by ward, and the `licenses` table lists businesses by ward.

The `land_use`, `census`, and `licenses` tables have been loaded for you.

#### Instructions 1/3         **35 XP**
* Merge `land_use` and `census` on the `ward` column. Merge the result of this with `licenses` on the `ward` column, using the suffix `_cen` for the left table and `_lic` for the right table. Save this to the variable `land_cen_lic`.


In [43]:
land_use = pd.read_pickle("../data/land_use.p")
census = pd.read_pickle("../data/census.p")
licenses = pd.read_pickle("../data/licenses.p")

# Merge land_use and census and merge result with licenses including suffixes
land_cen_lic = land_use.merge(census, on='ward') \
    .merge(licenses, on='ward', suffixes=('_cen', '_lic'))

land_cen_lic.head()

ward  residential  commercial  industrial  vacant  other  pop_2000  \
0    1           41           9           2       2     46     52951   
1    1           41           9           2       2     46     52951   
2    1           41           9           2       2     46     52951   
3    1           41           9           2       2     46     52951   
4    1           41           9           2       2     46     52951   

   pop_2010 change                  address_cen zip_cen account  aid  \
0     56149     6%  2765 WEST SAINT MARY STREET   60647   12024  NaN   
1     56149     6%  2765 WEST SAINT MARY STREET   60647   14446  743   
2     56149     6%  2765 WEST SAINT MARY STREET   60647   14624  775   
3     56149     6%  2765 WEST SAINT MARY STREET   60647   14987  NaN   
4     56149     6%  2765 WEST SAINT MARY STREET   60647   15642  814   

               business              address_lic zip_lic  
0   DIGILOG ELECTRONICS       1038 N ASHLAND AVE   60622  
1      EMPTY BOTTLE INC   1035 N WESTERN AVE 1ST   60622  
2  LITTLE MEL'S HOT DOG    2205 N CALIFORNIA AVE   60647  
3    MR. BROWN'S LOUNGE   2301 W CHICAGO AVE 1ST   60622  
4          Beat Kitchen  2000-2100 W DIVISION ST   60622

#### Instruction 2/3
* Group `land_cen_lic` by `ward`, `pop_2010` (the population in 2010), and `vacant`, then count the number of `accounts`. Save the results to `pop_vac_lic`.


In [46]:
# Merge land_use and census and merge result with licenses including suffixes
land_cen_lic = land_use.merge(census, on='ward') \
                    .merge(licenses, on='ward', suffixes=('_cen','_lic'))

# Group by ward, pop_2010, and vacant, then count the # of accounts
pop_vac_lic = land_cen_lic.groupby(["ward", "pop_2010", "vacant"],
                                   as_index=False).agg({'account':'count'})

pop_vac_lic.head()

ward  pop_2010  vacant  account
0    1     56149       2      253
1   10     51535      14      130
2   11     51497       5      201
3   12     52235       4      255
4   13     53722       1      101

#### Instruction 3/3

* Sort `pop_vac_lic` by `vacant`, `account`, and`pop_2010` in descending, ascending, and ascending order respectively. Save it as `sorted_pop_vac_lic`.


In [52]:
# Merge land_use and census and merge result with licenses including suffixes
land_cen_lic = land_use.merge(census, on='ward') \
                    .merge(licenses, on='ward', suffixes=('_cen','_lic'))

# Group by ward, pop_2010, and vacant, then count the # of accounts
pop_vac_lic = land_cen_lic.groupby(['ward','pop_2010','vacant'],
                                   as_index=False).agg({'account':'count'})

# Sort pop_vac_lic and print the results
sorted_pop_vac_lic = pop_vac_lic.sort_values(['vacant', 'account', 'pop_2010'],
                                             ascending=[False, True, True])

# Print the top few rows of sorted_pop_vac_lic
print(sorted_pop_vac_lic.head())

   ward  pop_2010  vacant  account
47    7     51581      19       80
12   20     52372      15      123
1    10     51535      14      130
16   24     54909      13       98
7    16     51954      13      156
